// electron-forge/dist/util/check-system.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _bluebird = require('bluebird');

var checkGitExists = function () {
  var _ref = (0, _bluebird.coroutine)( /*#__PURE__*/_regenerator2.default.mark(function _callee() {
    return _regenerator2.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            return _context.abrupt('return', new _promise2.default(function (resolve) {
              (0, _child_process.exec)('git --version', function (err) {
                if (err) return resolve(false);
                resolve(true);
              });
            }));

          case 1:
          case 'end':
            return _context.stop();
        }
      }
    }, _callee, this);
  }));

  return function checkGitExists() {
    return _ref.apply(this, arguments);
  };
}();

var checkNodeVersion = function () {
  var _ref2 = (0, _bluebird.coroutine)( /*#__PURE__*/_regenerator2.default.mark(function _callee2() {
    return _regenerator2.default.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            return _context2.abrupt('return', _promise2.default.resolve(_semver2.default.gt(process.versions.node, '6.0.0')));

          case 1:
          case 'end':
            return _context2.stop();
        }
      }
    }, _callee2, this);
  }));

  return function checkNodeVersion() {
    return _ref2.apply(this, arguments);
  };
}();

var checkPackageManagerVersion = function () {
  var _ref3 = (0, _bluebird.coroutine)( /*#__PURE__*/_regenerator2.default.mark(function _callee3(ora) {
    return _regenerator2.default.wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            return _context3.abrupt('return', (0, _yarnOrNpm.yarnOrNpmSpawn)(['--version']).then(function (version) {
              if ((0, _yarnOrNpm.hasYarn)()) {
                warnIfPackageManagerIsntAKnownGoodVersion('Yarn', version, YARN_WHITELISTED_VERSIONS, ora);
              } else {
                warnIfPackageManagerIsntAKnownGoodVersion('NPM', version, NPM_WHITELISTED_VERSIONS, ora);
              }

              return true;
            }));

          case 1:
          case 'end':
            return _context3.stop();
        }
      }
    }, _callee3, this);
  }));

  return function checkPackageManagerVersion(_x) {
    return _ref3.apply(this, arguments);
  };
}();

exports.isNightlyYarnVersion = isNightlyYarnVersion;

var _child_process = require('child_process');

var _semver = require('semver');

var _semver2 = _interopRequireDefault(_semver);

var _yarnOrNpm = require('./yarn-or-npm');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var NPM_WHITELISTED_VERSIONS = {
  all: '^3.0.0 || ^4.0.0 || ~5.1.0 || ~5.2.0 || >= 5.4.2',
  darwin: '>= 5.4.0',
  linux: '>= 5.4.0'
};
var YARN_WHITELISTED_VERSIONS = {
  all: '0.23.3 || 0.24.6 || >= 1.0.0',
  darwin: '0.27.5',
  linux: '0.27.5'
};

function isNightlyYarnVersion(version) {
  return (/((?:\d\.?)+)-\d.*/.test(version)
  );
}

function warnIfPackageManagerIsntAKnownGoodVersion(packageManager, version, whitelistedVersions, ora) {
  var osVersions = whitelistedVersions[process.platform];
  var versions = osVersions ? whitelistedVersions.all + ' || ' + osVersions : whitelistedVersions.all;
  var versionString = version.toString();
  if (isNightlyYarnVersion(versionString) || !_semver2.default.satisfies(versionString, versions)) {
    ora.warn('You are using ' + packageManager + ', but not a known good version.\n' + ('The known versions that work with Electron Forge are: ' + versions));
  }
}

exports.default = function () {
  var _ref4 = (0, _bluebird.coroutine)( /*#__PURE__*/_regenerator2.default.mark(function _callee4(ora) {
    return _regenerator2.default.wrap(function _callee4$(_context4) {
      while (1) {
        switch (_context4.prev = _context4.next) {
          case 0:
            _context4.next = 2;
            return _promise2.default.all([checkGitExists(ora), checkNodeVersion(ora), checkPackageManagerVersion(ora)]);

          case 2:
            _context4.t0 = function (check) {
              return check;
            };

            return _context4.abrupt('return', _context4.sent.every(_context4.t0));

          case 4:
          case 'end':
            return _context4.stop();
        }
      }
    }, _callee4, this);
  }));

  return function (_x2) {
    return _ref4.apply(this, arguments);
  };
}();
